src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -10,12 +17,112 @@ namespace DB
 
 namespace ErrorCodes
 {
+extern const int CANNOT_WRITE_TO_KAFKA;
 extern const int MISSING_ACKNOWLEDGEMENT;
 extern const int INVALID_CONFIG_PARAMETER;
+extern const int TYPE_MISMATCH;
+}
+
+namespace KafkaStream
+{
+ChunkPartitioner::ChunkPartitioner(ContextPtr context, const Block & header, const ASTPtr & partitioning_expr_ast) {
+    /// `InterpreterCreateQuery::handleExternalStreamCreation` ensures this
+    assert(partitioning_expr_ast);
+
+    ASTPtr query = partitioning_expr_ast;
+    auto syntax_result = TreeRewriter(context).analyze(query, header.getNamesAndTypesList());
+    partitioning_expr = ExpressionAnalyzer(query, syntax_result, context).getActions(false);
+
+    partitioning_key_column_name = partitioning_expr_ast->getColumnName();
+
+    if (auto * shard_func = partitioning_expr_ast->as<ASTFunction>())
+        if (shard_func->name == "rand" || shard_func->name == "RAND")
+        {
+            random_partitioning = true;
+            std::random_device r;
+            rand = std::minstd_rand(r());
+        }
+}
+
+BlocksWithShard ChunkPartitioner::partition(Block block, Int32 partition_cnt) const
+{
+    /// no topics have zero partitions
+    assert(partition_cnt > 0);
+    /// check the ctor for how this is guaruanteed
+    assert(partitioning_expr);
```

Comment:
Once we assert in ctor, we are good. So no need to assert here since it won't regret ?